# Loading Data


In [77]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [78]:
url ='https://raw.githubusercontent.com/Adu9o9/Heart-Disease-Prediction-ML/refs/heads/main/heart_disease_uci.csv'
df = pd.read_csv(url)
df

id  age     sex        dataset               cp  trestbps   chol    fbs  \
0      1   63    Male      Cleveland   typical angina     145.0  233.0   True   
1      2   67    Male      Cleveland     asymptomatic     160.0  286.0  False   
2      3   67    Male      Cleveland     asymptomatic     120.0  229.0  False   
3      4   37    Male      Cleveland      non-anginal     130.0  250.0  False   
4      5   41  Female      Cleveland  atypical angina     130.0  204.0  False   
..   ...  ...     ...            ...              ...       ...    ...    ...   
915  916   54  Female  VA Long Beach     asymptomatic     127.0  333.0   True   
916  917   62    Male  VA Long Beach   typical angina       NaN  139.0  False   
917  918   55    Male  VA Long Beach     asymptomatic     122.0  223.0   True   
918  919   58    Male  VA Long Beach     asymptomatic       NaN  385.0   True   
919  920   62    Male  VA Long Beach  atypical angina     120.0  254.0  False   

              restecg  thalch  exang  oldpeak        slope   ca  \
0      lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1      lv hypertrophy   108.0   True      1.5         flat  3.0   
2      lv hypertrophy   129.0   True      2.6         flat  2.0   
3              normal   187.0  False      3.5  downsloping  0.0   
4      lv hypertrophy   172.0  False      1.4    upsloping  0.0   
..                ...     ...    ...      ...          ...  ...   
915  st-t abnormality   154.0  False      0.0          NaN  NaN   
916  st-t abnormality     NaN    NaN      NaN          NaN  NaN   
917  st-t abnormality   100.0  False      0.0          NaN  NaN   
918    lv hypertrophy     NaN    NaN      NaN          NaN  NaN   
919    lv hypertrophy    93.0   True      0.0          NaN  NaN   

                  thal  num  
0         fixed defect    0  
1               normal    2  
2    reversable defect    1  
3               normal    0  
4               normal    0  
..                 ...  ...  
915                NaN    1  
916                NaN    0  
917       fixed defect    2  
918                NaN    0  
919                NaN    1  

[920 rows x 16 columns]

In [79]:
print("--- Initial Data (First 5 Rows) ---")
print(df.head())

--- Initial Data (First 5 Rows) ---
   id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num  
0       fixed defect    0  
1             normal    2  
2  reversable defect    1  
3             normal    0  


In [80]:
print("\n--- Data Info ---")
df.info()


--- Data Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        920 non-null    int64  
 1   age       920 non-null    int64  
 2   sex       920 non-null    object 
 3   dataset   920 non-null    object 
 4   cp        920 non-null    object 
 5   trestbps  861 non-null    float64
 6   chol      890 non-null    float64
 7   fbs       830 non-null    object 
 8   restecg   918 non-null    object 
 9   thalch    865 non-null    float64
 10  exang     865 non-null    object 
 11  oldpeak   858 non-null    float64
 12  slope     611 non-null    object 
 13  ca        309 non-null    float64
 14  thal      434 non-null    object 
 15  num       920 non-null    int64  
dtypes: float64(5), int64(3), object(8)
memory usage: 115.1+ KB


## Data Preprocessing


In [81]:
print("--- Null Values Before Cleaning ---")
print(df.isnull().sum())

--- Null Values Before Cleaning ---
id            0
age           0
sex           0
dataset       0
cp            0
trestbps     59
chol         30
fbs          90
restecg       2
thalch       55
exang        55
oldpeak      62
slope       309
ca          611
thal        486
num           0
dtype: int64


In [82]:
# Separate columns by data type
numerical_cols = df.select_dtypes(include=np.number).columns
categorical_cols = df.select_dtypes(include=['object']).columns

# Impute numerical columns with the median
for col in numerical_cols:
    df[col] = df[col].fillna(df[col].median())

# Impute categorical columns with the mode
for col in categorical_cols:
    df[col] = df[col].fillna(df[col].mode()[0])

/tmp/ipython-input-3281287293.py:11: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[col] = df[col].fillna(df[col].mode()[0])


In [83]:
print("\n--- Null Values After Cleaning ---")
print(df.isnull().sum())


--- Null Values After Cleaning ---
id          0
age         0
sex         0
dataset     0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalch      0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64


In [84]:
df.shape


(920, 16)

In [85]:
df.describe()

id         age    trestbps        chol      thalch     oldpeak  \
count  920.000000  920.000000  920.000000  920.000000  920.000000  920.000000   
mean   460.500000   53.510870  131.995652  199.908696  137.692391    0.853261   
std    265.725422    9.424685   18.451300  109.040171   25.145235    1.058049   
min      1.000000   28.000000    0.000000    0.000000   60.000000   -2.600000   
25%    230.750000   47.000000  120.000000  177.750000  120.000000    0.000000   
50%    460.500000   54.000000  130.000000  223.000000  140.000000    0.500000   
75%    690.250000   60.000000  140.000000  267.000000  156.000000    1.500000   
max    920.000000   77.000000  200.000000  603.000000  202.000000    6.200000   

               ca         num  
count  920.000000  920.000000  
mean     0.227174    0.995652  
std      0.628936    1.142693  
min      0.000000    0.000000  
25%      0.000000    0.000000  
50%      0.000000    1.000000  
75%      0.000000    2.000000  
max      3.000000    4.000000

In [86]:
# We convert 'num' (0,1,2,3,4) into 'disease' (0 or 1)
df['disease'] = df['num'].apply(lambda x: 1 if x > 0 else 0)


## Data Preparation

In [87]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score

In [88]:
# We use our new 'disease' column as the target.0 for no disease and 1 for some sort of heart disease
y = df['disease']
X = df.drop(columns=['num', 'disease'])

# Step 1: Automatically find all columns that contain text (dtype 'object')
categorical_cols = X.select_dtypes(include=['object']).columns
print(f"Found text-based columns to convert: {list(categorical_cols)}")

# Step 2: Convert all of these text columns into numbers at once
X = pd.get_dummies(X, columns=categorical_cols, drop_first=True)

print("\nData is now fully prepared and numeric.")
X.info() # This will confirm all columns are numbers


Found text-based columns to convert: ['sex', 'dataset', 'cp', 'restecg', 'slope', 'thal']

Data is now fully prepared and numeric.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id                        920 non-null    int64  
 1   age                       920 non-null    int64  
 2   trestbps                  920 non-null    float64
 3   chol                      920 non-null    float64
 4   fbs                       920 non-null    bool   
 5   thalch                    920 non-null    float64
 6   exang                     920 non-null    bool   
 7   oldpeak                   920 non-null    float64
 8   ca                        920 non-null    float64
 9   sex_Male                  920 non-null    bool   
 10  dataset_Hungary           920 non-null    bool   
 11  dataset_Switzerland       920 non-null    bo

## Data Splitting

In [89]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, classification_report
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

## Model Building & Evaluation
### Logistic Regression

In [90]:
lr = LogisticRegression(max_iter=3900, random_state=42)
lr.fit(X_train, y_train)
y_lr_pred = lr.predict(X_test)
lr_accuracy = accuracy_score(y_test, y_lr_pred)
lr_precision = precision_score(y_test, y_lr_pred)

### RandomForest


In [91]:
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)
y_rf_pred = rf.predict(X_test)
rf_accuracy = accuracy_score(y_test, y_rf_pred)
rf_precision = precision_score(y_test, y_rf_pred, average='weighted') # Use 'weighted' average

In [92]:
from sklearn.metrics import classification_report

print("--- Multiclass Classification Report ---")
print(classification_report(y_test, y_rf_pred))

--- Multiclass Classification Report ---
              precision    recall  f1-score   support

           0       0.89      0.80      0.85        82
           1       0.85      0.92      0.89       102

    accuracy                           0.87       184
   macro avg       0.87      0.86      0.87       184
weighted avg       0.87      0.87      0.87       184



## Model Comparison Table

In [93]:
df_models = pd.DataFrame({
    'Method': ['Logistic Regression', 'Random Forest'],
    'Test Accuracy': [lr_accuracy, rf_accuracy],
    'Test Weighted Precision': [lr_precision, rf_precision]
})

In [94]:
print("\n--- Model Comparison Table ---")
print(df_models.reset_index(drop=True))


--- Model Comparison Table ---
                Method  Test Accuracy  Test Weighted Precision
0  Logistic Regression       0.847826                 0.849057
1        Random Forest       0.869565                 0.871189


### CONCLUSION
"The objective of this project was to build a model to predict the presence of heart disease. Two models were trained and evaluated: Logistic Regression and Random Forest.

Based on the results, the Random Forest model was the superior performer, achieving an overall accuracy of 87.0% and a weighted precision of 87.1% on the test set. This indicates a strong ability to correctly flag individuals who have the condition, which is a critical requirement for a medical diagnostic model. The Logistic Regression model was also effective but slightly less so, with an accuracy of 84.8%.

Therefore, the final recommendation is the Random Forest model for this prediction task."